File: 2023/AOC2023.Tests/Day01Tests.cs
using System.Text.RegularExpressions;
using Shouldly;

namespace AOC2023.Tests;

public class Day01Tests : TestBase
{
    protected override void SolvePart1_Sample()
    {
        // Arrange
        var expectedCalibrationSum = 142;
        var calibrationValues = get_sample()
            .Select(GetNumberFromFirstAndLastDigitsInString).ToList();

        // Act
        var calibrationSum = calibrationValues.Sum();
        
        // Assert
        calibrationSum.ShouldBe(expectedCalibrationSum);
    }
    
    protected override void SolvePart1_Actual()
    {
        // Arrange
        var expectedCalibrationSum = 54601;
        var calibrationValues = get_input()
            .Select(GetNumberFromFirstAndLastDigitsInString).ToList();

        // Act
        var calibrationSum = calibrationValues.Sum();
        
        // Assert
        calibrationSum.ShouldBe(expectedCalibrationSum);
    }

    private int GetNumberFromFirstAndLastDigitsInString(string input)
    {
        var justNumbers = Regex.Replace(input, @"[^\d]", "");
        var charArray = new char[]{ justNumbers.First(), justNumbers.Last()};
        var firstAndLast = new string(charArray);
        return int.Parse(firstAndLast);
    }
    
    protected override void SolvePart2_Sample()
    {
        // Arrange
        var expectedCalibrationSum = 281;
        var calibrationValues = get_sample(2)
            .Select(GetNumberFromFirstAndLastDigitsOrSpellingsInString).ToList();

        // Act
        var calibrationSum = calibrationValues.Sum();
        
        // Assert
        calibrationSum.ShouldBe(expectedCalibrationSum);
    }
    
    protected override void SolvePart2_Actual()
    {
        // Arrange
        var expectedCalibrationSum = 54078;
        var calibrationValues = get_input()
            .Select(GetNumberFromFirstAndLastDigitsOrSpellingsInString).ToList();

        // Act
        var calibrationSum = calibrationValues.Sum();
        
        // Assert
        calibrationSum.ShouldBe(expectedCalibrationSum);
    }
    
    private int GetNumberFromFirstAndLastDigitsOrSpellingsInString(string input)
    {
        var spellings = _spelledNumberMap.Select(x => x.spelling);
        var numRegex = new Regex($@"(\d|{string.Join("|", spellings)})");
        var matches = new List<Match>();
        matches.Add(numRegex.Match(input));
        while (matches.Last().Success) {
            matches.Add(numRegex.Match(input, matches.Last().Index + 1)); 
        }
        matches.RemoveAt(matches.Count - 1);
        var firstMatch = matches.First().Value;
        var lastMatch = matches.Last().Value;
        var firstNumber = _spelledNumberMap.FirstOrDefault(x => x.spelling == firstMatch).value;
        firstNumber = firstNumber == 0 ? int.Parse(firstMatch) : firstNumber;
        var lastNumber = _spelledNumberMap.FirstOrDefault(x => x.spelling == lastMatch).value;
        lastNumber = lastNumber == 0 ? int.Parse(lastMatch) : lastNumber;
        return firstNumber * 10 + lastNumber;
    }

    private List<(string spelling, int value)> _spelledNumberMap = new List<(string spelling, int value)>
    {
        ("one", 1),
        ("two", 2),
        ("three", 3),
        ("four", 4),
        ("five", 5),
        ("six", 6),
        ("seven", 7),
        ("eight", 8),
        ("nine", 9)
    };
}

File: 2023/kotlin/aoc-2023/src/test/kotlin/Day01Tests.kt
import kotlin.test.assertEquals

class Day01Tests : TestBase() {
    override fun solvePart1Sample() {

        val expectedCalibrationSum = 142
        val calibrationValues = getSample()
            .map { getNumberFromFirstAndLastDigitsInString(it) }
        val calibrationSum = calibrationValues.sum()

        assertEquals(expectedCalibrationSum, calibrationSum)
    }

    override fun solvePart1Actual() {

        val expectedCalibrationSum = 54601
        val calibrationValues = getInput()
            .map { getNumberFromFirstAndLastDigitsInString(it) }
        val calibrationSum = calibrationValues.sum()

        assertEquals(expectedCalibrationSum, calibrationSum)
    }

    private fun getNumberFromFirstAndLastDigitsInString(input: String) : Int {
        val regex = Regex("""\D""")
        val justNumbers = input.replace(regex, "")
        val charArray = charArrayOf(justNumbers.first(), justNumbers.last())
        val firstAndLast = String(charArray)
        return firstAndLast.toInt()
    }
    override fun solvePart2Sample() {
        val expectedCalibrationSum = 281
        val calibrationValues = getSample(2)
            .map { getNumberFromFirstAndLastLDigitsOrSpellingsInString(it) }
        val calibrationSum = calibrationValues.sum()
        assertEquals(expectedCalibrationSum, calibrationSum)
    }

    override fun solvePart2Actual() {
        val expectedCalibrationSum = 54078
        val calibrationValues = getInput()
            .map { getNumberFromFirstAndLastLDigitsOrSpellingsInString(it) }
        val calibrationSum = calibrationValues.sum()
        assertEquals(expectedCalibrationSum, calibrationSum)
    }

    private fun getNumberFromFirstAndLastLDigitsOrSpellingsInString(input: String): Int {
        val spellings = SpelledNumberMap.entries.map { e -> e.toString() }
        val numRegex = Regex("(\\d|${spellings.joinToString("|")})")
        val matches = mutableListOf<MatchResult?>()
        matches.add(numRegex.find(input, 0))
        while (matches.last() != null){
            matches.add(numRegex.find(input, matches.last()!!.range.first + 1))
        }
        matches.removeLast()
        val firstMatch = matches.first()!!.value
        val lastMatch = matches.last()!!.value
        var firstNumber = SpelledNumberMap.entries.find { m -> m.name == firstMatch }?.value?:0
        if (firstNumber == 0)
            firstNumber = firstMatch.toInt()
        var lastNumber = SpelledNumberMap.entries.find { m -> m.name == lastMatch }?.value?:0
        if (lastNumber == 0)
            lastNumber = lastMatch.toInt()
        return firstNumber * 10 + lastNumber
    }

    enum class SpelledNumberMap(val value: Int){
        one(1),
        two(2),
        three(3),
        four(4),
        five(5),
        six(6),
        seven(7),
        eight(8),
        nine(9)
    }
}